pandas/io/excel/_openpyxl.py

```diff
@@ -451,3 +462,277 @@ def write_cells(self, cells, sheet_name=None, startrow=0, startcol=0,
+        return load_workbook(filepath_or_buffer, data_only=True)
+
+    @property
+    def sheet_names(self):
+        return self.book.sheetnames
+
+    @staticmethod
+    def _handle_usecols(frame, usecols):
+        column_names = frame.columns.values
+        if usecols:
+            _validate_usecols_arg(usecols)
+            usecols = sorted(usecols)
+            if any(isinstance(i, str) for i in usecols):
+                _validate_usecols_names(usecols, column_names)
+                frame = frame[usecols]
+            else:
+                frame = frame.iloc[:, usecols]
+        return frame
+
+    def _handle_sheet_name(self, sheet_name):
+        """Handle the sheet_name keyword."""
+        # Keep sheetname to maintain backwards compatibility.
+        if isinstance(sheet_name, list):
+            sheets = sheet_name
+        elif sheet_name is None:
+            sheets = self.sheet_names
+        else:
+            sheets = [sheet_name]
+        return sheets
+
+    @staticmethod
+    def _handle_header_keywords(data, header, skiprows, index_col):
+        """Handle keywords relating to header parsing."""
+        # forward fill and pull out names for MultiIndex column
+        header_names = None
+        if header is not None and is_list_like(header):
+            header_names = []
+            control_row = [True] * len(data[0])
+
+            for row in header:
+                if is_integer(skiprows):
+                    row += skiprows
+
+                data[row], control_row = _fill_mi_header(data[row],
+                                                         control_row)
+
+                if index_col is not None:
+                    header_name, _ = _pop_header_name(data[row], index_col)
+                    header_names.append(header_name)
+        return header_names
+
+    @staticmethod
+    def _handle_convert_float(series, convert_float):
+        """Handle the convert_float keyword."""
+        # attempt to convert object columns to integer. Only because this
+        # is implicitly done when reading and excel file with xlrd, that
+        # behaviour is replicated here.
+
+        if series.dtype == object:
+            try:
+                series = series.astype('int64')
+            except (ValueError, TypeError):
+                try:
+                    series = series.astype('float64')
+                except (ValueError, TypeError):
+                    return series
+        elif (convert_float
+                and series.dtype >= float
+                and all(series % 1 == 0)):
+            series = series.astype('int64')
+        elif not convert_float:
+            if series.dtype >= int:
+                series = series.astype('float64')
+        return series
+
+    @staticmethod
+    def _handle_index_col(frame, index_col):
+        column_names = frame.columns.values
+        if index_col is None:
+            return frame
+        if is_list_like(index_col):
```

Comment:
This is all just documented as only accept an int or list of int. Are you actually seeing str anywhere? 

If not you should be able to greatly simplify by doing something like

```python
frame = frame.set_index(frame.columns.take(index_col).tolist())
```